### DISCUSSION TITLE
How to communicate between agent context and application context

### DISCUSSION THREAD

**Discussion by wallezhang** [open-telemetry/opentelemetry-java-instrumentation#7485] [Q&A] [answered]
The situation is: I create a new context in my application by `Context.current()`, and set this new context into a Map for cross thread communication. While read the context from Map by the agent instrumentation in another thread, a `ClassCastException` is thrown: `class io.opentelemetry.javaagent.instrumentation.opentelemetryapi.context.AgentContextWrapper cannot be cast to class io.opentelemetry.javaagent.shaded.io.opentelemetry.context.Context`.

I wonder if I should not put Context directly into the map, but replace it with the text converted by `TextMapPropagator`. Or is there any other good way?

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @wallezhang ,

The javaagent should propagate context automatically over threads in most cases (e.g. `Context.current()` called in a `Runnable` should return the context captured at the time when the runnable was submitted to an executor). Does that not work for you?

> I create a new context in my application by `Context.current()`, and set this new context into a Map for cross thread communication. While read the context from Map by the agent instrumentation in another thread, a `ClassCastException` is thrown: `class io.opentelemetry.javaagent.instrumentation.opentelemetryapi.context.AgentContextWrapper cannot be cast to class io.opentelemetry.javaagent.shaded.io.opentelemetry.context.Context`.

The application and the agent are using two different copies of the OpenTelemetry API. The agent one is shaded (thus the `io.opentelemetry.javaagent.shaded` package prefix). The application OTel API is being bridged by the `opentelemetry-api` instrumentation; it effectively implements the application's interfaces and forwards all the calls to the (shaded) agent API. 
The `AgentContextWrapper` class that you're seeing is a part of the `opentelemetry-api` instrumentation; it implements the application `Context` interface and bridges all the calls over to the agent `Context` (the shaded one).

When you put the value of `Context.current()` into some kind of shared map (on the application side), you actually put an instance of `AgentContextWrapper` there. So, when you retrieve it from the javaaagent code space, the object not implement the agent's `Context` interface (and thus the class cast error). You could use `AgentContextStorage.getAgentContext()` to extract the actual agent context from an `AgentContextWrapper`. In general, I'd recommend avoiding this kind of app-agent communication if possible (just try to put all the parts either on the application side or the agent side, not both of them at once), as coordinating it gets rather bothersome.

**Reply by wallezhang** [reply]
Thanks for your answer. I understand the reason of this mistake through your answer. As you say, if I use `AgentContextStorage.getAgentContext()` to extract the actual agent context, I should import this dependency in my agent extension:
```xml
<dependency>
  <groupId>io.opentelemetry.javaagent.instrumentation</groupId>
  <artifactId>opentelemetry-javaagent-opentelemetry-api-1.0</artifactId>
</dependency>
```

Then I should follow the following steps:

1. Put the context(`AgentContextWrapper`) to a standalone field on the application side
2. Retrieve the context(`AgentContextWrapper`) from the standalone field on agent side and cast to `AgentContextWrapper` object
3. Use `AgentContextStorage.getAgentContext()` to extract the actual agent context
4. Use the actual agent context in the instrumentation module

Is my understanding above is correct?

**Reply by mateuszrzeszutek** [reply]
> I should import this dependency in my agent extension:

Yeah, exactly that one 👍 

> Is my understanding above is correct?

Yep, that sounds good!

**Reply by wallezhang** [reply]
Thanks 😄

**Reply by wallezhang** [reply]
When I tried to follow the steps, I found that when I use the `AgentContextStorage.getAgentContext()` extract the agent context,  here's another problem, the parameter signature of this method is `application.io.opentelemetry.context.Context`, I can't find this class in my custom agent extension.

<img width="930" alt="image" src="https://user-images.githubusercontent.com/33749/210323751-116ba527-16b0-49c5-afc2-9edae0abd27b.png">

**Reply by mateuszrzeszutek** [reply]
Yeah, it's a hack that we do to be able to reference both application & agent OTel APIs. We have a [shaded OTel API module](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/opentelemetry-api-shaded-for-instrumenting/build.gradle.kts) in this project that's not published (and it won't be; it's an internal implementation detail), and that's referenced as a [`compileOnly` dependency](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/build.gradle.kts#L17-L35) in the `opentelemetry-api` instrumentation module. You might have to duplicate that setup in your own project.

(The above is one of the reasons why I suggested to keep all the code either in the application or the agent space if possible; mixing it makes things kinda complicated)

**Reply by wallezhang** [reply]
Yeah, I know this hack. The question is: why is the `application.io.opentelemetry` not modified back to `io.opentelemetry` while being published? I unpack the `opentelemetry-javaagent.jar` and found the class is already refers to `io.opentelemetry` in `opentelemetry-api-1.0` module.

**Reply by mateuszrzeszutek** [reply]
Ah, good point. We publish unshaded instrumentations to make referencing/using them easier - if we shaded the `application...` package, we'd have to shade the agent copy of the API too, otherwise they'd conflict. And in general we don't want our users to worry about shading - in most cases, custom extension developers don't really have to think about it, they can reference unshaded APIs and the agent will handle the shading/translation in the runtime.

**Reply by wallezhang** [reply]
Oh, thanks, I see why published dependency was not shaded. I'll just have to go and make a duplicate of shaded OTel API module then, or put all the logic into the agent extension.

**Reply by wallezhang** [reply]
😢 There would be another problem here, I make a copy of the `opentelemetry-api-shaded` module, I need to shade `application.io.opentelemetry` to `io.opentelemetry` and then shade the `io.opentelemetry` used in the agent extension to `io.opentelemetry.shaded` when agent extension is compiled. But then, when the extension is loaded via `-Dotel.javaagent.extension`, it will replace `io.opentelemetry` in the class file with `io.opentelemetry.shaded` again by the `io.opentelemetry.javaagent.tooling. RemappingUrlConnection` class.

If I don't shade `application.io.opentelemetry`, extension will refer to `application.io.opentelemetry...` at runtime, this couldn't work. Maybe also shade `application.io.opentelemetry` in `RemappingUrlConnection` class?

**Reply by mateuszrzeszutek** [reply]
> If I don't shade `application.io.opentelemetry`, extension will refer to `application.io.opentelemetry...` at runtime, this couldn't work. Maybe also shade `application.io.opentelemetry` in `RemappingUrlConnection` class?

Hm, that's a fair point. Can you create a bug for this?

**Reply by wallezhang** [reply]
Ok, I'll create a bug and glad to fix it.
